Agent: fix agent quit when promotion failed

jinuxstyle opened the pull request:
Fixes #1191.

When promoting a node to manager failed, it should and could stay
working as agent role, although it's role is changed to manager
in advance. The ideal solution should change its role back to agent
as well, but it will get more parts involved and require a larger
change. It's a long term item. For now, keep the node running is
better than quiting. It's actually the same behavior as if the
m.Run fails.

Signed-off-by: Jin Xu <jinuxstyle@hotmail.com>

aaronlehmann commented:
ping @tonistiigi @LK4D4

tonistiigi commented:
>  It's actually the same behavior as if the
m.Run fails.

Indeed, but it doesn't seem correct. @LK4D4 How are we handling random errors from `m.Run()`. Nothing should be switching the role to agent then.

Promotion problems are not the only errors that can happen there and for other errors we don't want to wait for demotion. Also if this doesn't happen on promotion but for example on restoring with old settings then failing a node is the correct behavior. If this is about detecting nodes without proper listening address, maybe we can validate this before starting manager(or return a typed error) but even then we should have an endpoint for the manager to refuse its promotion because waiting for a demotion is only one possible way how this error can be resolved. For example if the address is in use correct behavior is to either error or repeatedly try again.

jinuxstyle commented:
@tonistiigi Insightful. Let's see @LK4D4 's opinion on how to handle potential errors for m.Run.

stevvooe commented on agent/node.go:
```diff
@@ -627,7 +627,19 @@ func (n *Node) runManager(ctx context.Context, securityConfig *ca.SecurityConfig
 		})
 		if err != nil {
-			return err
+			logrus.Errorf("error starting manager: %v", err)
```
`log.G(ctx).WithError(err).Error("error starting manager")`

jinuxstyle replied:
OK.

stevvooe commented on agent/node.go:
```diff
@@ -627,7 +627,19 @@ func (n *Node) runManager(ctx context.Context, securityConfig *ca.SecurityConfig
+			// TODO(jinuxstyle): change role back to agent if promotion
+			// failed in a self-healing way.
+			n.waitRole(ctx, ca.AgentRole)
+			if ctx.Err() != nil {
```
Why do the context check here?

jinuxstyle replied:
The context might be cancelled when waiting the role change.

stevvooe replied:
That should be handled by line 677. Perhaps, that needs to move. Really, this should use `<-ctx.Done()`. I don't think it is threadsafe to call `ctx.Err()` before `ctx.Done()`.

jinuxstyle replied:
I agree calling ctx.Err() after ctx.Done() is a good practice. But is it really not threadsafe if it's called before ctx.Done(). The context.Cancel() will set context.Err to non-nil and close context.Done immediately after that.

stevvooe replied:
In practice, it probably is, but that should not be relied upon. The interface indicates that `ctx.Done()` should be closed before the value is valid.

jinuxstyle commented:
Updated PR per @tonistiigi 's comments with one more commit which handles the potential errors of m.Run(). Also added check of whether a node's initial role is manager or not to decide whether to quit or wait.

tonistiigi commented:
@LK4D4 Maybe we should make a separate PR first that handles the manager returning out of `m.Run()` (if you see it as an issue).
@jinuxstyle The rest of this pr confuses me a bit because why would a node that has been promoted to a manager(initialRole=worker) behave any differently then a node that has been promoted to a manager and then restarted(initialRole=manager).

LK4D4 commented:
@tonistiigi I think a manager will try to restart while node has manager role. There is no error handling from Run, though. Also, I don't see any way for a manager to exit Run apart from demotion, raft node failure or Stop call. Not sure how bad this stuff is.

tonistiigi commented:
> ... raft node failure ...

@LK4D4 If raft node fails we should restart the manager, not wait for a demotion, right?

LK4D4 commented:
@tonistiigi we should and if I'm reading code correctly - we do.

jinuxstyle commented:
I will strip the second commit out and put it into a separate PR per @tonistiigi 's suggestion.

jinuxstyle commented:
@tonistiigi Regarding the initialRole that confused you, I think the difference is that when promoting the node whose initial role is worker, it's not a member of the raft, while the node whose initial role is manager when it's restarted, it was already a member of the raft. So the failure of running manager part when restarting a node is critical but promoting is not.

jinuxstyle commented:
PR updated: stripped out the second commit and slightly improved the first commit per review comments.

jinuxstyle commented:
Opened issue #1409 for handling the error of m.Run.

stevvooe commented on agent/node.go:
```diff
@@ -600,7 +601,26 @@ func (n *Node) runManager(ctx context.Context, securityConfig *ca.SecurityConfig
+			}
+
+			n.waitRole(ctx, ca.AgentRole)
+			if ctx.Err() != nil {
```
Again, this is not the correct way to use `ctx.Err()`. You need to listen on `ctx.Done()` first.

The correct approach would be to return an error from `waitRole`.

jinuxstyle replied:
I agree with the approach you suggested. Indeed, the waitRole() used to return error until commit 294ade7bd976796. I don't if there is any strong reason for that. Ping @LK4D4 for comment.

aluzzardi commented:
@jinuxstyle - are you planning on continuing this PR?

jinuxstyle commented:
Yes if review can progress. And if you think it's better to solve it by core maintainers due to it's difficulty, feel free to take it over.

aluzzardi commented:
Thanks @jinuxstyle 

ping @LK4D4 @tonistiigi
